rfd/0113-moderated-file-transfers.md

```diff
@@ -0,0 +1,176 @@
+
+---
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 113 - Web Moderated Session File Transfer
+
+## Required approvers
+
+Engineering: @zmb3 @xacrimon @jakule
+
+## What
+
+Approval process to allow file transfers during a web moderated session
+
+## Why
+
+Users have no ability to transfer files if their role requires a moderated session for the specified server, even if they are in an active and "valid" session currently. Originally, we would allow file transfers regardless of moderated sessions which, while good for UX, goes against the purpose of moderated sessions in the first place. We have disabled the functionality but would like to reintroduce it in a secure, moderated, and auditable way. 
+
+## Scope
+
+This RFD will cover how to implement for web SCP specifically, but may be expanded to SFTP for tsh, although my research has been mostly for web scp. If at anytime there is ambiguity in this RFD for "which type of file transfers does this refer to", default to web scp only. 
+
+## The current issue
+
+When a user initiates a file transfer request (HTTP request and `tcp scp`), the system establishes a new non-interactive session to run the "exec" command. When we attempt to [open the session](https://github.com/gravitational/teleport/blob/64b10f1ccbeeab63c4ce91f5f11fdd74b42448d8/lib/srv/sess.go#L277) we have a check named `checkIfStart` which will pull any policies that require moderation from the current identity, and return if they've been fulfilled or not. This will always fail as this is a "new" session, created just a few lines before the check itself. Therefore, we need a way to let the server know that this request has been moderator approved.
+
+### Details
+
+#### The FileTransferRequest 
+
+The general idea of the solution is to provide a way for users to "request-a-request" that can be approved by moderators. An example of a file transfer specific struct could look like
```

Comment:
Will this new request only be used for moderated sessions or will it apply any time there is a web based file transfer? 

If it won't apply every time, how will you detect whether or not you need to use it?

Reply:
Only for moderated sessions. If it isn't a moderated session, no extra parameters are needed to be sent and will just pass the request without them added. I speak about [it here](https://github.com/gravitational/teleport/blob/ccdd62aa2d94e25dd896591db4057fec73359962/rfd/0113-moderated-file-transfers.md#updated-file-transfer-api-handler) but I see its a bit unclear now. Will update.

The web console will determine if we need to start this new process, and the api handler knowing when to add the environment variables based on the existence of the optional parameters sent. Sort of similar to how we [issue certs when a webauthn response is included](https://github.com/gravitational/teleport/pull/22864/files#diff-ad4065e07d2573425dc5f7460e26dd324ffd893845ae1d3fe4b997113aba90cbR133), we would be able to set the environment variables needed for `OpenExecSession`. 

Also, `OpenExecSession` knows if the soon-to-be-opened session is moderated or not, so we can have a check "if moderated, see if they've sent through a commandRequestID".